pkg/manager/impl/reservation_manager.go

```diff
@@ -0,0 +1,188 @@
+) interfaces.ReservationManager {
+	systemMetrics := reservationMetrics{
+		scope: reservationScope,
+		reservationAcquiredViaCreate: labeled.NewCounter(
+			"reservation_acquired_via_create",
+			"Number of times a reservation was acquired via create",
+			reservationScope),
+		reservationAcquiredViaUpdate: labeled.NewCounter(
+			"reservation_acquired_via_update",
+			"Number of times a reservation was acquired via update",
+			reservationScope),
+		reservationAlreadyInProgress: labeled.NewCounter(
+			"reservation_already_in_progress",
+			"Number of times we try of acquire a reservation but the reservation is in progress",
+			reservationScope,
+		),
+		makeReservationFailure: labeled.NewCounter(
+			"make_reservation_failure",
+			"Number of times we failed to make reservation",
+			reservationScope,
+		),
+		getTagFailure: labeled.NewCounter(
+			"get_tag_failure",
+			"Number of times we failed to get tag",
+			reservationScope,
+		),
+	}
+
+	return &reservationManager{
+		repo:               repo,
+		reservationTimeout: reservationTimeout,
+		now:                nowFunc,
+		systemMetrics:      systemMetrics,
+	}
+}
+
+func (r *reservationManager) GetOrReserveArtifact(ctx context.Context, request *datacatalog.GetOrReserveArtifactRequest) (*datacatalog.GetOrReserveArtifactResponse, error) {
+	tagKey := transformers.ToTagKey(request.DatasetId, request.TagName)
+	tag, err := r.repo.TagRepo().Get(ctx, tagKey)
+	if err != nil {
+		if errors2.IsDoesNotExistError(err) {
+			// Tag does not exist yet, let's reserve a spot to work on
+			// generating the artifact.
+			status, err := r.makeReservation(ctx, request)
+			if err != nil {
+				r.systemMetrics.makeReservationFailure.Inc(ctx)
+				return nil, err
+			}
+
+			return &datacatalog.GetOrReserveArtifactResponse{
+				Value: &datacatalog.GetOrReserveArtifactResponse_ReservationStatus{
+					ReservationStatus: &status,
+				},
+			}, nil
+		}
+		logger.Errorf(ctx, "Failed retrieve tag: %+v, err: %v", tagKey, err)
+		r.systemMetrics.getTagFailure.Inc(ctx)
+		return nil, err
+	}
+
+	artifact, err := transformers.FromArtifactModel(tag.Artifact)
+	if err != nil {
+		return nil, err
+	}
+
+	return &datacatalog.GetOrReserveArtifactResponse{
+		Value: &datacatalog.GetOrReserveArtifactResponse_Artifact{
+			Artifact: artifact,
+		},
+	}, nil
+}
+
+func (r *reservationManager) makeReservation(ctx context.Context, request *datacatalog.GetOrReserveArtifactRequest) (datacatalog.ReservationStatus, error) {
+	repo := r.repo.ReservationRepo()
+	reservationKey := transformers.ToReservationKey(*request.DatasetId, request.TagName)
+	rsv, err := repo.Get(ctx, reservationKey)
+
+	if err != nil {
+		if errors2.IsDoesNotExistError(err) {
+			// Reservation does not exist yet so let's create one
+			err := repo.Create(ctx, models.Reservation{
+				ReservationKey: reservationKey,
+				OwnerID:        request.OwnerId,
+				ExpireAt:       r.now().Add(r.reservationTimeout),
+			})
+
+			if err != nil {
```

Comment:
what about the race condition case where two processes attempt to create the same reservation after asserting it doesn't exist? will the primary key guarantee a failure due to duplication for the second call?

Reply:
Yeah the primary key will ensure the second attempt fails with already exists.